### PR TITLE
fix: prevent appendWithCap from splitting surrogate pairs

### LINKED ISSUE #636: Bug: Invalid UTF-8 in heartbeat_runs breaks list endpoint with 500
## Summary

`GET /api/companies/:companyId/heartbeat-runs` returns HTTP 500 with:

```
PostgresError: invalid byte sequence for encoding "UTF8": 0xe2 0x22 0x2c
```

The byte sequence `0xe2 0x22 0x2c` indicates a truncated 3-byte UTF-8 character: `0xe2` is a leading byte expecting two continuation bytes (0x80–0xBF), but instead is followed by ASCII `"` (0x22) and `,` (0x2c). This means a multi-byte character was split during truncation.

## Root Cause

`appendWithCap()` in `packages/adapter-utils/src/server-utils.ts:67-70`:

```typescript
export function appendWithCap(prev: string, chunk: string, cap = MAX_CAPTURE_BYTES) {
  const combined = prev + chunk;
  return combined.length > cap ? combined.slice(combined.length - cap) : combined;
}
```

The constants are named `MAX_CAPTURE_BYTES` and `MAX_EXCERPT_BYTES`, implying byte-level limits, but:
- `string.length` counts **UTF-16 code units**, not bytes
- `.slice()` operates on code units and can split **surrogate pairs** (emoji, supplementary characters)
- When surrogate pairs are split, encoding the resulting string to UTF-8 produces invalid byte sequences

This function is used by `appendExcerpt()` in `server/src/services/heartbeat.ts` to truncate `stdout_excerpt` and `stderr_excerpt` before storing them in PostgreSQL.

Additionally, `summarizedHeartbeatRunResultJson` (heartbeat.ts:49-79) uses PostgreSQL's `left()` to truncate JSONB text extractions to 500 chars — if the source JSONB was already stored with corrupted encoding, `left()` can propagate or expose the invalid sequences.

Once a single corrupted row exists, **the entire list endpoint is broken** — every request to list heartbeat runs returns 500 because the query touches the corrupted row.

## Steps to Reproduce

1. Create an agent that produces output containing emoji or supplementary Unicode characters (e.g., `console.log("Processing… ✅ Done 🎉")`)
2. Let the agent run until `stdout_excerpt` exceeds `MAX_EXCERPT_BYTES` (32 KB)
3. `appendWithCap` truncates the string, potentially splitting a multi-byte character
4. The corrupted string is stored in `heartbeat_runs.stdout_excerpt`
5. Any subsequent `GET /api/companies/:companyId/heartbeat-runs` returns 500

## Observed Behavior

- The 500 error repeats on every poll of the heartbeat-runs endpoint (the UI polls it periodically)
- Individual queries selecting specific columns (e.g., `SELECT id, status, created_at`) succeed
- Querying with `length(stdout_excerpt)` also succeeds — the corruption only surfaces when PostgreSQL reads the actual text bytes for serialization

## Suggested Fix

Option A — Use `Buffer.byteLength` for accurate byte counting and truncate on character boundaries:

```typescript
export function appendWithCap(prev: string, chunk: string, cap = MAX_CAPTURE_BYTES) {
  let combined = prev + chunk;
  while (Buffer.byteLength(combined, 'utf8') > cap) {
    combined = combined.slice(1);
  }
  return combined;
}
```

Option B — Sanitize before storing by replacing broken surrogates:

```typescript
function sanitizeUtf8(str: string): string {
  // Remove lone surrogates that would produce invalid UTF-8
  return str.replace(/[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?<![\uD800-\uDBFF])[\uDC00-\uDFFF]/g, '\uFFFD');
}
```

Also consider adding a `CATCH` wrapper in `summarizedHeartbeatRunResultJson` so one corrupted row doesn't break the entire list endpoint.

## Data Recovery

Existing corrupted rows can be fixed with:

```sql
UPDATE heartbeat_runs
SET stdout_excerpt = regexp_replace(stdout_excerpt, E'[\\x80-\\xFF]', '', 'g')
WHERE company_id = '<affected_company_id>';
```

Or selectively null them out:

```sql
UPDATE heartbeat_runs
SET stdout_excerpt = NULL
WHERE company_id = '<affected_company_id>'
  AND stdout_excerpt IS NOT NULL
  AND NOT (stdout_excerpt::bytea IS NOT NULL);
```

## Introduced In

- `631c859` (2026-02-18) — *Move adapter implementations into shared workspace packages* — introduced `appendWithCap()`
- `c9c75bb` (2026-02-17) — *Implement agent runtime services and WebSocket realtime* — introduced `appendExcerpt()` calling `appendWithCap`
- `92aef9b` (2026-03-10) — *Slim heartbeat run list payloads* — introduced `summarizedHeartbeatRunResultJson` SQL truncation

## Environment

- Paperclip server (local embedded-postgres mode)
- postgres.js v3.4.8
- PostgreSQL embedded (UTF-8 encoding)

### PR DESCRIPTION
**Pull request by vkartaviy**
## Summary

- Fix `appendWithCap()` in `packages/adapter-utils/src/server-utils.ts` to not produce invalid UTF-8 when truncating strings containing emoji or supplementary Unicode characters
- `string.slice()` operates on UTF-16 code units and can split a surrogate pair, producing a lone surrogate that becomes an invalid UTF-8 byte sequence when stored in PostgreSQL
- After slicing, drop a leading lone low surrogate (U+DC00–U+DFFF) and a trailing lone high surrogate (U+D800–U+DBFF) so the result is always valid Unicode

## Changes

1. **`packages/adapter-utils/src/server-utils.ts`** — after `slice()`, strip lone low surrogate at start and lone high surrogate at end
2. **`packages/adapter-utils/src/server-utils.test.ts`** — 12 unit tests covering surrogate pair handling
3. **`packages/adapter-utils/vitest.config.ts`** + **`vitest.config.ts`** — register `adapter-utils` in the vitest project list

## Test plan

- [x] Unit tests pass (`npx vitest run packages/adapter-utils/src/server-utils.test.ts` — 12/12)
  - Basic truncation (under cap, at cap, over cap)
  - Empty prev / empty chunk
  - Lone low surrogate at start after slicing → dropped
  - Lone high surrogate at end (partial write) → dropped
  - Cap of 1 with surrogate pair → empty string
  - Valid pairs preserved when boundary is clean
  - Consecutive surrogate pairs at various boundaries
  - Default cap (MAX_CAPTURE_BYTES)
- [ ] Verify heartbeat-runs list endpoint no longer returns 500 when agent output contains emoji near the truncation boundary

Fixes #636

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/adapter-utils/src/server-utils.ts [hunk 1]
```diff
@@ -66,7 +66,13 @@ export function parseJson(value: string): Record<string, unknown> | null {
 
 export function appendWithCap(prev: string, chunk: string, cap = MAX_CAPTURE_BYTES) {
   const combined = prev + chunk;
-  return combined.length > cap ? combined.slice(combined.length - cap) : combined;
+  if (combined.length <= cap) return combined;
+  let result = combined.slice(combined.length - cap);
+  // If slice cut through a surrogate pair, drop the leading lone surrogate
+  if (result.length > 0 && result.charCodeAt(0) >= 0xdc00 && result.charCodeAt(0) <= 0xdfff) {
+    result = result.slice(1);
+  }
```
[the remark this answers is not in the document]
Comment (vkartaviy) [reply packages/adapter-utils/src/server-utils.ts:74]: Fixed [`022fe92` (this PR)](https://github.com/paperclipai/paperclip/pull/639/commits/022fe92a67992a70ff8d95a37286f0fcea767e4c)

**Comment by vkartaviy** [comment]
Unit tests are added ☝🏼

**Review by SollanSystems — APPROVED** [review]
LGTM. Correct surrogate pair handling — detects lone high/low surrogates at slice boundaries and drops them. Thorough test coverage including edge cases (consecutive pairs, cap=1, empty strings). Clean fix.
